mautic/mautic
# A/B test enhancements

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [Y]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [Y] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes https://github.com/mautic/mautic/issues/1752



#### Description:

_This is temporarily created against the 4.4 even though this won't be pushed to M4 as there won't be any more feature releases. However, we have an agreement with @npracht that his team will prepare this feature for M5 if Acquia provides the backbone of it. Community FTW!_

This feature will let users to send the A/B test to a small subset of contacts, decides which version is the winner based on configured criteria and then the rest of the segment(s) will receive the winner variant.

An example of how it would look:

Start with a segment of 1,000 contacts

1.) Set up an A/B Test with three different emails, each of a weight of 10%

Email A - send to 100 people
Email B - send to 100 people
Email C - send to 100 people

2.) Create a 'winning criteria' i.e. email open rate

Email A - Open rate is 25%
Email B - Open rate is 15%
Email C - Open rate is 10%

3.) The winning email is sent to the rest of the contacts after a predetermined amount of time

Email A is sent to remaining 70% of segment 3 hours later

New command for sending winner variant:
mautic: email:sendwinner - searches for all emails with variant settings for automatic sending the winner variant, can be used by a cron job
mautic: email:sendwinner --id=101 - works for a single email only

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Crete Email, enable AB test in settings, set total weight for the AB test and predetermined amount of hours for sending Winner variant to remaining contacts
3. Add A/B test variant, set weight
4. Send email, variants are sent using total weight for the AB test
5. Run mautic: email:sendwinner command, if the command is run after predetermined amount of hours, email is sent to remaining contacts, winner variant is converted to parent variant, weight of the winner variant is set to 100 and other variants are unpublished. Test it also when a winner variant is already the parent one.
6. Verify that it works also with email variants created before this change
7. Verify that Page variants are still working the same, part of the functionality is shared


#### List backwards compatibility breaks:
1. Variant criteria is now saved in parent variant, instead of each variant child. Should still work with already created variants in old way.
